contents/functionalProgramming/intro.md

```diff
@@ -0,0 +1,132 @@
+# An Introduction to Functional Programming
+
+Authors: [Phang Chun Rong](https://github.com/crphang)
+
+## What is Functional Programming
+
+Functional programming is a programming paradigm that treats computation as the evaluation of mathematical function and avoids mutating state and variables. Unlike [imperative programming](https://en.wikipedia.org/wiki/Imperative_programming) languages like C, functional programming languages are [declarative](https://en.wikipedia.org/wiki/Declarative_programming). Even though this demands a shift in mindset from most programmers who are used to imperative form of programming, functional programming paradigm is still [increasingly popular](https://blog.appdynamics.com/engineering/the-most-popular-programming-languages-for-2017/) due to its [advantages](#advantages-of-functional-programming).
+
+## Functional Programming Languages
+
+Functional programming is simply a paradigm and needs to be implemented by programming languages. Below are various [languages](https://en.wikipedia.org/wiki/List_of_programming_languages_by_type#Functional_languages) that have explicit support for the functional programming paradigm such as:
+
+- Haskell
+- Clojure
+- Elixir
+- Elm
+- Scala
+- ML (Meta Language) family of languages
+
+While functional programming can be implemented by languages like Java, the languages listed above encourage the functional programming paradigm such as [pure functions](#pure-functions) or even enforce them in the case of Haskell.
+
+## Concepts in Functional Programming
+
+#### Pure Functions
+
+Pure functions is one of the concepts in functional programming. To call a function `pure`, it needs to satisfy two conditions:
+
+1. [Idempotence](https://en.wikipedia.org/wiki/Idempotence) - The function should return the same output for the same input when invoked arbitrarily many times.
+1. No side effects - The function should not cause side effects like mutating global state.
+
+To illustrate an example of a pure function. Consider this simple example in Python:
+
+```python
+# Pure Function
+def add(x, y):
+    return x + y
+```
+
+Contrast this with an impure function:
+
+```python
+# Mutate state
+y=0
+def addXToY(x):
+    global y
+    y = y + x
+    return y
+```
+We see that the impure function violates both conditions. Running `addXToY` multiple times with the same input X would give different value.
+
+#### Immutability
+
+Another concept in functional programming is immutability. Immutability prevents an object's state to be change after it is created. When we assign `x=1`, it does not make mathematical sense to have this expression `x:=x+1`.
```

Comment:
> it does not make mathematical sense to have this expression `x:=x+1`

While this is true, I'm not sure that equates to an explanation or even a justification of immutability. The sentence  sounds like this equation doesn't make sense, mathematically. i.e., the syntax is nonsensical